src/internal/m365/exchange/events_restore.go

```diff
@@ -102,8 +108,115 @@ func (h eventRestoreHandler) restore(
 		return nil, clues.Stack(err)
 	}
 
+	// Fix up event instances in case we have a recurring event
+	err = h.updateRecurringEvents(ctx, userID, destinationID, ptr.Val(item.GetId()), event)
+	if err != nil {
+		return nil, clues.Stack(err)
+	}
+
 	info := api.EventInfo(event)
 	info.Size = int64(len(body))
 
 	return info, nil
 }
+
+func (h eventRestoreHandler) updateRecurringEvents(
+	ctx context.Context,
+	userID, containerID, eventID string,
+	event models.Eventable,
+) error {
+	if event.GetRecurrence() == nil {
+		return nil
+	}
+
+	// Cancellations and exceptions are currently in additional data
+	// but will get their own fields once the beta API lands and
+	// should be moved then
+	cancelledOccurrences := event.GetAdditionalData()["cancelledOccurrences"]
+	exceptionOccurrences := event.GetAdditionalData()["exceptionOccurrences"]
+
+	// OPTIMIZATION: Group instances whose dates are close by
+	if cancelledOccurrences != nil {
+		co, ok := cancelledOccurrences.([]interface{})
+		if !ok {
+			return clues.New("converting cancelledOccurrences to []interface{}").
+				With("type", fmt.Sprintf("%T", cancelledOccurrences))
+		}
+
+		for _, insti := range co {
+			inst, ok := insti.(*string)
+			if !ok {
+				return clues.New("converting canceled instance to *string").
+					With("type", fmt.Sprintf("%T", insti))
+			}
+
+			splits := strings.Split(ptr.Val(inst), ".")
+			startStr := splits[len(splits)-1]
+
+			start, err := time.Parse(string(dttm.DateOnly), startStr)
+			if err != nil {
+				return clues.Wrap(err, "parsing cancelled event date")
+			}
+
+			endStr := dttm.FormatTo(start.Add(24*time.Hour), dttm.DateOnly)
+
+			evts, err := h.ac.GetItemInstances(ctx, userID, eventID, startStr, endStr)
+			if err != nil {
+				return clues.Wrap(err, "getting instances")
+			}
+
+			if len(evts) != 1 {
+				return clues.New("invalid number of instances").
+					With("count", len(evts))
```

Comment:
"count" is overly generic and could be misunderstood (in the best case) or clobber existing values (in the worst).  Please namespace your keys.
```suggestion
					With("event_instance_count", len(evts))
```